main.go

```diff
@@ -30,90 +25,111 @@ func main() {
 		return
 	}
 
-	// Read the plugin output from protoc.
-	req, err := codegenerator.ParseRequest(os.Stdin)
-	if err != nil {
-		log.Fatal(err)
+	protogen.Options{}.Run(func(gen *protogen.Plugin) error {
+		// Parse the parameters handed to the plugin.
+		param := parseParams(gen.Request.GetParameter())
+
+		// Iterate over each file passed to the plugin.
+		for _, f := range gen.Files {
+			if !f.Generate {
+				continue
+			}
+
+			// Extract the RPC call godoc from the proto file.
+			godoc := extractComments(f)
+
+			// Generate stubs either for mobile or for JS.
+			if param["js_stubs"] == "1" {
+				genJSStubs(gen, f, param)
+			} else {
+				genMobileStubs(gen, f, param, godoc)
+			}
+
+			// Finally, with the service definitions successfully
+			// created, create the in-memory grpc definitions if
+			// requested.
+			if param["mem_rpc"] == "1" {
+				genMemRPC(gen, f, param)
+			}
+		}
+
+		return nil
+	})
+}
+
+// parseParams parses any parameters handed to the plugin.
+func parseParams(parameter string) map[string]string {
+	param := make(map[string]string)
+	if parameter == "" {
+		return param
 	}
 
-	// Load the parsed request into a descriptor registry.
-	reg := descriptor.NewRegistry()
-	if err := reg.Load(req); err != nil {
-		log.Fatal(err)
+	for _, p := range strings.Split(parameter, ",") {
+		if i := strings.Index(p, "="); i < 0 {
+			param[p] = ""
+		} else {
+			param[p[0:i]] = p[i+1:]
+		}
 	}
 
-	// Parse the parameters handed to the plugin.
-	parameter := req.GetParameter()
-	param := split(parameter, ",")
+	return param
+}
+
+// extractComments extracts the RPC call godoc from the proto file.
+func extractComments(file *protogen.File) map[string]string {
+	locations := file.Desc.SourceLocations()
 
-	// Extract the RPC call godoc from the proto.
 	godoc := make(map[string]string)
-	for _, f := range req.GetProtoFile() {
-		fd := &generator.FileDescriptor{
-			FileDescriptorProto: f,
+	for i := 0; i < locations.Len(); i++ {
+		loc := locations.Get(i)
+
+		if loc.LeadingComments == "" {
+			continue
 		}
-		for _, loc := range fd.GetSourceCodeInfo().GetLocation() {
-			if loc.LeadingComments == nil {
-				continue
+		c := loc.LeadingComments
+
+		// Find the first newline. The actual comment will start
+		// following this.
+		i := 0
+		for j := range c {
```

Comment:
nit: we could use `strings.Index` instead (here and below).

Reply:
good idea!